Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1

```diff
@@ -150,18 +169,49 @@ function Set-CurrentAzureRMSubscription {
     $null = Select-AzureRMSubscription -SubscriptionId $SubscriptionId @additional
 }
 
-function Format-Splat {
+function Set-UserAgent {
     [CmdletBinding()]
-    param([Parameter(Mandatory = $true)][hashtable]$Hashtable)
-
-    # Collect the parameters (names and values) in an array.
-    $parameters = foreach ($key in $Hashtable.Keys) {
-        $value = $Hashtable[$key]
-        # If the value is a bool, format the parameter as a switch (ending with ':').
-        if ($value -is [bool]) { "-$($key):" } else { "-$key" }
-        $value
+    param()
+
+    $collectionUri = Get-VstsTaskVariable -Name System.TeamFoundationCollectionUri -Require
+    $collectionId = Get-VstsTaskVariable -Name System.CollectionId -Require
+    $hostType = Get-VstsTaskVariable -Name System.HostType -Require
+    $serverString = "TFS"
+    if ($collectionUri.ToLower().Contains("visualstudio.com".ToLower())) {
+        $serverString = "VSTS"
     }
+    
+    $userAgent = [string]::Empty
+    if ($hostType -ieq "build") {
+		$definitionId = Get-VstsTaskVariable -Name System.DefinitionId -Require
+		$buildId = Get-VstsTaskVariable -Name Build.BuildId -Require
+		$userAgent = $serverString + "_" + $collectionId + "_" + "build" + "_" + $definitionId + "_" + $buildId
+    } elseif ($hostType -ieq "release") {
+        $definitionId = Get-VstsTaskVariable -Name Release.DefinitionId -Require
+		$releaseId = Get-VstsTaskVariable -Name Release.ReleaseId -Require
+		$environmentId = Get-VstsTaskVariable -Name Release.EnvironmentId -Require
+		$attemptNumber = Get-VstsTaskVariable -Name Release.AttemptNumber -Require
+		$userAgent = $serverString + "_" + $collectionId + "_" + "release" + "_" + $definitionId + "_" + $releaseId + "_" + $environmentId + "_" + $attemptNumber
+    } else {
+	    return
+	}
+	
+    Set-UserAgent_Core -UserAgent $userAgent
+}
 
-    $OFS = " "
-    "$parameters" # String join the array.
+function Set-UserAgent_Core {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        [string]$UserAgent)
+
+    Trace-VstsEnteringInvocation $MyInvocation
+    try {
+        [Microsoft.Azure.Common.Authentication.AzureSession]::ClientFactory.AddUserAgent($UserAgent)
+    } catch {
+	    Write-Host "##vso[task.logissue type=warning;TaskName=SetUserAgent]"
```

Comment:
What does this do? `TaskName=SetUserAgent`

I would recommend just logging to verbose stream. Otherwise the customer ends up with a warning issue attached to their build for an issue that A) doesn't affect them and B) they can't fix.

You may want to log the exception message to the verbose stream, so that in the event we see the logs, it includes something useful to diagnose why it's not added.